packages/react/src/theme-extensions/with-default-color-scheme.ts

```diff
@@ -0,0 +1,35 @@
+import defaultTheme from "@chakra-ui/theme"
+import { ThemingProps } from "@chakra-ui/system"
+import { Dict, isObject } from "@chakra-ui/utils"
+import { mergeThemeOverride, ThemeExtension } from "../extend-theme"
+
+export function withDefaultColorScheme({
+  colorScheme,
+  components = defaultTheme.components,
+}: {
+  colorScheme: ThemingProps["colorScheme"]
+  components?: string[] | Dict
+}): ThemeExtension {
+  return (theme) => {
+    let names = Object.keys(theme.components || {})
+
+    if (Array.isArray(components)) {
+      names = components
+    } else if (isObject(components)) {
+      names = Object.keys(components)
+    }
```

Comment:
It seems like this logic isn't quite what we'd want.

We set `names` to `theme.components`, and then we immediately replace that with `components` or `Object.keys(components`). That would be fine except `components` defaults to `defaultTheme.components`, so there's no way to have it use `theme.components` for the `names`. Which imo would be a major use case, so if a theme with custom component themes is passed, they're automatically included.

Reply:
Same for the other functions with the same setup.

Reply:
This code defaults to all components in the theme. If the argument `components` is defined and is an array or object, it uses those components. 

In this test case the Alert component is not explicitly passed, but receives the colorScheme:

https://github.com/chakra-ui/chakra-ui/blob/1a1e980ba9b84bd50e56d167c71312f7804e7562/packages/react/tests/with-default-color-scheme.test.tsx#L5-L21

Reply:
Right! But what I'm saying is that it defaults to the wrong `theme`'s components. Like the last `expect` here fails:

```ts
const themeWithCustomComponent = extendTheme({
  components: {
    MyCustomComponent: {
      defaultProps: {
        colorScheme: 'purple',
      },
    },
  },
})
const theme = extendTheme(
  {
    colors: {
      brand: {
        500: '#b4d455',
      },
    },
  },
  withDefaultColorScheme({colorScheme: 'brand'}),
  themeWithCustomComponent,
)

expect(theme.components.Alert.defaultProps.colorScheme).toBe('brand')
expect(theme.components.MyCustomComponent.defaultProps.colorScheme).toBe(
  'brand',
)
```

Reply:
`names` is initially correct in getting the names from `theme.components` rather than `defaultTheme.components`, but the `if`/`else if` then immediately overwrite that with `defaultTheme.components` since that's the default value.

Reply:
Oh, good catch 👀 💪 I did not see the default arg there. Thanks for pointing that out!

Looking forward to improve the API and code with you 😊 🎉

Reply:
I removed the default value for the parameter `components` 👍

Reply:
And I added your mentioned test case (I had to reverse the order of overrides due `pipe`)